FragmentDirectiveGenerator: Selecting the first two or three words of the first sentence in a paragraph results in the whole page highlighting

hortont424 opened the pull request:
#### 0b92e6db304e4326e077b979552ef80ba56d5a1a
<pre>
FragmentDirectiveGenerator: Selecting the first two or three words of the first sentence in a paragraph results in the whole page highlighting
<a href="https://bugs.webkit.org/show_bug.cgi?id=278257">https://bugs.webkit.org/show_bug.cgi?id=278257</a>
<a href="https://rdar.apple.com/133772459">rdar://133772459</a>

Reviewed by Wenson Hsieh and Megan Gardner.

The bug in the title results from our unconditional generation of a prefix (or suffix)
fragment component in the case where there is no possible prefix or suffix.

Avoid generating empty components, and to do so, reorganize the generation code a bit.

Add a test for this case (and make FragmentDirectiveGenerator trivially testable).

* LayoutTests/http/tests/scroll-to-text-fragment/generation-does-not-emit-empty-prefix-expected.txt: Added.
* LayoutTests/http/tests/scroll-to-text-fragment/generation-does-not-emit-empty-prefix.html: Added.
* Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
(WebCore::FragmentDirectiveGenerator::generateFragmentDirective):
* Source/WebCore/dom/FragmentDirectiveGenerator.h:
(WebCore::FragmentDirectiveGenerator::urlWithFragment const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::textFragmentDirectiveForRange):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/282375@main">https://commits.webkit.org/282375@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/211dcdaf8f7ac52db853d88a783f29e114d6c523

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62996 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42352 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15592 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67017 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13600 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65116 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50039 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13884 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9380 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66065 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/39351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54554 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/31456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/36040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11888 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12476 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/57584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12218 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68712 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6942 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/58086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6974 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54615 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13971 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5787 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/38172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/39252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/40363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
```diff
@@ -102,38 +102,49 @@ void FragmentDirectiveGenerator::generateFragmentDirective(const SimpleRange& te
+    };
+
+    if (textFromRange.length() >= maximumInlineStringLength) {
+        startText = encodeComponent(nextWordsFromPosition(3, visibleStartPosition));
```
should the `3` be extracted to a named constant ?

megangardner replied:
3 is also elsewhere in the file, so change those to this constant as well, please!

hortont424 committed:
FragmentDirectiveGenerator: Selecting the first two or three words of the first sentence in a paragraph results in the whole page highlighting
https://bugs.webkit.org/show_bug.cgi?id=278257
rdar://133772459

Reviewed by Wenson Hsieh and Megan Gardner.

The bug in the title results from our unconditional generation of a prefix (or suffix)
fragment component in the case where there is no possible prefix or suffix.

Avoid generating empty components, and to do so, reorganize the generation code a bit.

Add a test for this case (and make FragmentDirectiveGenerator trivially testable).

* LayoutTests/http/tests/scroll-to-text-fragment/generation-does-not-emit-empty-prefix-expected.txt: Added.
* LayoutTests/http/tests/scroll-to-text-fragment/generation-does-not-emit-empty-prefix.html: Added.
* Source/WebCore/dom/FragmentDirectiveGenerator.cpp:
(WebCore::FragmentDirectiveGenerator::generateFragmentDirective):
* Source/WebCore/dom/FragmentDirectiveGenerator.h:
(WebCore::FragmentDirectiveGenerator::urlWithFragment const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::textFragmentDirectiveForRange):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/282375@main
